library/lua/script-data.lua

```diff
@@ -0,0 +1,344 @@
+    -- (if there's nothing to load, we'll give the data table from the new entry as the persistent table to use)
+    entry = create_loaded_data_entry(id, "world")
+    path = get_save_dir("world") .. "/" .. id .. ".json"
+  elseif type == "global" then
+    entry = create_loaded_data_entry(id, "global")
+    path = get_save_dir("global") .. "/" .. id .. ".json"
+  end
+
+
+  -- Load file (if it exists), and update the entry's stored persistent table to contain the loaded data
+  if dfhack.filesystem.isfile(path) then
+    entry.data = json.decode_file(path)
+  end
+
+  return entry.data
+end
+
+---------------------------------------------------------------------
+-- GLOBAL FUNCTIONS
+
+-- Returns a table whose content is saved into the world save. If there's saved data under the given ID in the world, that data is loaded into the table.
+-- Autosave is an optional flag which causes the data to be saved automatically during autosaves/quicksaves if enabled (and is enabled by default)
+function load_world(id, autosave)
+  local persistent_table = load_data(id, "world")
+
+  -- Set the autosave flag (if provided). Otherwise default to on
+  set_autosave(id, autosave or true)
+
+  return persistent_table
+end
+
+-- Returns a table whose content is available regardless of currently loaded save. If there's saved data under the given ID, that data is loaded into the table.
+function load_global(id)
+  local persistent_table = load_data(id, "global")
+
+  return persistent_table
+end
+
+-- Trigger the contents of the persistent table of a given ID to be saved to a file.
+function save(id)
+  -- Trigger presave? TODO
+
+  local path
+  if get_loaded_data(id).type == "global" then
+    ensure_save_dir_exists("global")
+    path = get_save_dir("global") .. "/" .. id .. ".json"
+  else -- assume world
+    --[[
+    -- Safety check for arena mode
+    -- Arena mode saves are the only(?) worlds that can trigger world loaded/unloaded without already having an existing save file
+    -- DFhack's dfhack.getSavePath() will provide a path for that save game's folder even if it doesn't actually exist!
+
+    -- Check the save folder actually exists before attempting to save - if it doesn't, abort!
+    update_save_path()
+
+    if not dfhack.filesystem.isdir(recorded_save_path) then
+      return false
+    end
+
+    -- Otherwise, continue as usual
+    ]]
+    --^ Old safety check for before quit_without_saving was added to work around it (the check didn't really work properly anyway)
+    ensure_save_dir_exists("world")
+    path = get_save_dir("world") .. "/" .. id .. ".json"
+  end
+
+  json.encode_file(get_loaded_data(id).data, path)
+end
+
+-- Trigger saves for all persistent tables which have the autosave flag currently set
+-- This is naturally triggered by this script whenever an autosave / quicksave is detected
+function save_autosave()
+  for id, entry in pairs(loaded_data) do
+    if entry.autosave == true then
```

Comment:
A minor stylistic point: I typically suggest replacing `if x == true` and `if x == false` with `if x` and `if not x`, respectively. `nil` and `false` are the only things that implicitly evaluate to false in Lua when treated as a boolean, so you could also replace `if x == nil` with `if not x` in most cases (although the fact that 0 evaluates to true could be confusing, particularly to anyone less familiar with Lua implementation details, so I don't tend to push this one as hard).

Reply:
I've been trying to intentionally structure checks like this nowadays to improve readability, though at the slight cost of elegance (and I believe it's safe to do in this check because autosave can only ever be `true` or `false`, never `nil`). I originally learned lua - and how to use DFhack especially - by reading through scripts made by other people, and I personally find that making the conditions obvious means it's easier to tell what the checks mean to somebody reading the script who's learning lua, or transitioning from another language.